File: kotlin/app/src/main/kotlin/radixulous/app/layer_links.kt
package radixulous.app.opusmanager
import radixulous.app.structure.OpusTree
import radixulous.app.opusmanager.BeatKey
import radixulous.app.opusmanager.OpusEvent
import radixulous.app.opusmanager.FlagLayer

open class LinksLayer : FlagLayer {
    constructor() : super()
    var linked_beat_map: HashMap<BeatKey, BeatKey> = HashMap<BeatKey, BeatKey>()
    var inv_linked_beat_map: HashMap<BeatKey, MutableList<BeatKey>> = HashMap<BeatKey, MutableList<BeatKey>>()

    fun unlink_beat(beat_key: BeatKey) {
        if (! this.linked_beat_map.containsKey(beat_key)) {
           return
        }

        var target_key = this.linked_beat_map.get(beat_key)
        var beats: MutableList<BeatKey> = this.inv_linked_beat_map.get(target_key)!!
        beats.remove(beat_key)
        if (beats.size == 0) {
            this.inv_linked_beat_map.remove(target_key)
        }
        this.linked_beat_map.remove(beat_key)
    }

    fun clear_links_to_beat(beat_key: BeatKey) {
        if (! this.inv_linked_beat_map.containsKey(beat_key)) {
            return
        }
        var links = this.inv_linked_beat_map.get(beat_key)!!
        for (link_key in links) {
            this.linked_beat_map.remove(link_key)
        }
        this.inv_linked_beat_map.remove(beat_key)
    }

    fun link_beats(beat_key: BeatKey, target: BeatKey) {
        // Remove any existing link
        this.unlink_beat(beat_key)
        // Replace existing tree with a copy of the target
        this.overwrite_beat(beat_key, target)

        this.linked_beat_map.put(beat_key, target)
        if (! this.inv_linked_beat_map.containsKey(target)) {
            this.inv_linked_beat_map.put(target, mutableListOf())
        }
        this.inv_linked_beat_map[target]!!.add(beat_key)
    }
    private fun get_all_linked(beat_key: BeatKey): List<BeatKey> {
        var output: MutableList<BeatKey> = mutableListOf()
        if (this.inv_linked_beat_map.containsKey(beat_key)) {
            output.add(beat_key)
            for (linked_key in this.inv_linked_beat_map[beat_key]!!) {
                output.add(linked_key)
            }
        } else if (this.linked_beat_map.contains(beat_key)) {
            var target_key = this.linked_beat_map[beat_key]!!
            output.add(target_key)
            for (linked_key in this.inv_linked_beat_map[target_key]!!) {
                output.add(linked_key)
            }
        } else {
            output.add(beat_key)
        }

        return output
    }

    open override fun insert_after(beat_key: BeatKey, position: List<Int>) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.insert_after(linked_key, position)
        }
    }
    open override fun remove(beat_key: BeatKey, position: List<Int>) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.remove(linked_key, position)
        }
    }
    open override fun set_percussion_event(beat_key: BeatKey, position: List<Int>) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.set_percussion_event(linked_key, position)
        }
    }
    open override fun set_event(beat_key: BeatKey, position: List<Int>, value: Int, relative: Boolean) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.set_event(linked_key, position, value, relative)
        }
    }
    open override fun split_tree(beat_key: BeatKey, position: List<Int>, splits: Int) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.split_tree(linked_key, position, splits)
        }
    }
    open override fun unset(beat_key: BeatKey, position: List<Int>) {
        for (linked_key in this.get_all_linked(beat_key)) {
            super.unset(linked_key, position)
        }
    }

    /////////

    private fun remap_links(remap_hook: (beat_key: BeatKey, args: List<Int>) -> BeatKey?, args: List<Int>) {
        var new_link_map = HashMap<BeatKey, BeatKey>()
        this.inv_linked_beat_map.clear()
        for (beat in this.linked_beat_map.keys) {
            var target = this.linked_beat_map.get(beat)!!
            var new_beat = remap_hook(beat, args)
            var new_target = remap_hook(target, args)
            if (new_beat == null || new_target == null) {
                continue
            }
            new_link_map.put(new_beat, new_target)

            if (! this.inv_linked_beat_map.containsKey(new_target)) {
                this.inv_linked_beat_map.put(new_target, mutableListOf())
            }
            this.inv_linked_beat_map.get(new_target)!!.add(new_beat)
        }
        this.linked_beat_map = new_link_map
    }

    open override fun change_line_channel(old_channel: Int, line_offset: Int, new_channel: Int) {
        super.change_line_channel(old_channel, line_offset, new_channel)
        var new_offset = this.channel_trees[new_channel].size - 1
        this.remap_links(this::rh_change_line_channel, listOf(old_channel, line_offset, new_channel, new_offset))
    }
    private fun rh_change_line_channel(beat_key: BeatKey, args: List<Int>): BeatKey? {
        var old_channel = args[0]
        var line_offset = args[1]
        var new_channel = args[2]
        var new_offset = args[3]

        var new_beat = beat_key
        if (beat_key.channel == old_channel) {
            if (beat_key.line_offset == line_offset) {
                new_beat = BeatKey(new_channel, new_offset, beat_key.beat)
            } else if (beat_key.line_offset > line_offset) {
                new_beat = BeatKey(beat_key.channel, beat_key.line_offset - 1, beat_key.beat)
            }
        }
        return new_beat
    }

    open override fun move_line(channel: Int, old_index: Int, new_index: Int) {
        super.move_line(channel, old_index, new_index)

        this.remap_links(this::rh_move_line, listOf(channel, old_index, new_index))
    }

    private fun rh_move_line(beat: BeatKey, args: List<Int>): BeatKey? {
        var channel = args[0]
        var old_index = args[1]
        var new_index = args[2]

        var new_beat = beat
        if (beat.channel == channel) {
            if (beat.line_offset == old_index) {
                new_beat = BeatKey(beat.channel, new_index, beat.beat)
            } else if (old_index < beat.line_offset && beat.line_offset < new_index) {
                new_beat = BeatKey(beat.channel, beat.line_offset - 1, beat.beat)
            }
        }
        return new_beat
    }

    open override fun insert_beat(index: Int?) {
        super.insert_beat(index)

        var index_remap: Int
        if (index == null) {
            index_remap = this.opus_beat_count
        } else {
            index_remap = index
        }
        this.remap_links(this::rh_insert_beat, listOf(index_remap))
    }

    private fun rh_insert_beat(beat: BeatKey, args: List<Int>): BeatKey? {
        var index = args[0]
        var new_beat: BeatKey
        if (beat.beat >= index) {
            new_beat = BeatKey(beat.channel, beat.line_offset, beat.beat + 1)
        } else {
            new_beat = beat
        }
        return new_beat
    }

    open override fun remove_beat(index: Int?) {
        super.remove_beat(index)
        var adj_index: Int
        if (index == null) {
            adj_index = -1
        } else {
            adj_index = index
        }
        this.remap_links(this::rh_remove_beat, listOf(adj_index))
    }

    private fun rh_remove_beat(beat: BeatKey, args: List<Int>): BeatKey? {
        var index = args[0]
        var new_beat: BeatKey

        if (beat.beat >= index) {
            new_beat = BeatKey(beat.channel, beat.line_offset, beat.beat - 1)
        } else {
            new_beat = beat
        }

        return new_beat
    }

    open override fun remove_channel(channel: Int) {
        super.remove_channel(channel)
        this.remap_links(this::rh_remove_channel, listOf(channel))
    }

    private fun rh_remove_channel(beat: BeatKey, args: List<Int>): BeatKey? {
        if (beat.channel == args[0]) {
            return null
        } else {
            return beat
        }
    }

    open override fun remove_line(channel: Int, line_offset: Int?) {
        super.remove_line(channel, line_offset)
    }
    private fun rh_remove_line(beat: BeatKey, args: List<Int>): BeatKey? {
        var channel = args[0]
        var line_offset = args[1]
        var new_beat: BeatKey? = beat
        if (beat.channel == channel) {
            if (beat.line_offset == line_offset) {
                new_beat = null
            } else if (beat.line_offset > line_offset) {
                new_beat = BeatKey(channel, line_offset - 1, beat.beat)
            }
        }
        return new_beat
    }

    open override fun swap_channels(channel_a: Int, channel_b: Int) {
        super.swap_channels(channel_a, channel_b)
        this.remap_links(this::rh_swap_channels, listOf(channel_a, channel_b))
    }
    private fun rh_swap_channels(beat: BeatKey, args: List<Int>): BeatKey? {
        if (beat.channel == args[0]) {
            return BeatKey(args[1], beat.line_offset, beat.beat)
        } else if (beat.channel == args[1]) {
            return BeatKey(args[0], beat.line_offset, beat.beat)
        } else {
            return beat
        }
    }

    // TODO
    open public fun load_folder(path: String) { }
    open public fun save(path: String? = null) { }
}

File: src/opusmanager/layer_links.py
from __future__ import annotations
import os
import json

from typing import Optional, Dict, List, Tuple, TypeAlias

from .miditree import MIDITreeEvent
from .layer_base import OpusManagerBase, BeatKey

class LinksLayer(OpusManagerBase):
    """Adds linked beats."""
    def __init__(self):
        super().__init__()
        self.linked_beat_map = {}
        self.inv_linked_beat_map = {}

    def unlink_beat(self, beat_key: BeatKey) -> None:
        """
            Removes the link from this beat to another.
            Leaves the beat unchanged.
        """
        if beat_key not in self.linked_beat_map:
            return

        channel, line, beat = beat_key

        target_key = self.linked_beat_map[beat_key]
        self.inv_linked_beat_map[target_key].remove(beat_key)
        if not self.inv_linked_beat_map[target_key]:
            del self.inv_linked_beat_map[target_key]

        del self.linked_beat_map[beat_key]

    def clear_links_to_beat(self, target: BeatKey) -> None:
        """Remove all links pointing to the specific beat"""
        if target not in self.inv_linked_beat_map:
            return

        while self.inv_linked_beat_map[target]:
            linked_key = self.inv_linked_beat_map[target].pop()
            del self.linked_beat_map[linked_key]
        del self.inv_linked_beat_map[target]

    def link_beats(self, beat: BeatKey, target: BeatKey) -> None:
        """
            Overwrites *beat* with a copy of *target*,
            then creates a link between the two such that
            any changes made to one are made to the other.
        """
        # Remove any existing link
        self.unlink_beat(beat)

        # Replace existing tree with a copy of the target
        self.overwrite_beat(beat, target)

        self.linked_beat_map[beat] = target
        if target not in self.inv_linked_beat_map:
            self.inv_linked_beat_map[target] = set()
        self.inv_linked_beat_map[target].add(beat)


    def _get_all_linked(self, beat_key: BeatKey) -> List[BeatKey]:
        """
            Get a list of all positions of beats that will be affected
            by changes made to the one at the given position (ie, linked)
        """
        if beat_key in self.inv_linked_beat_map:
            positions = [beat_key]
            for linked_key in self.inv_linked_beat_map[beat_key]:
                positions.append(linked_key)
        elif beat_key in self.linked_beat_map:
            target_key = self.linked_beat_map[beat_key]
            positions = [target_key]
            for linked_key in self.inv_linked_beat_map[target_key]:
                positions.append(linked_key)
        else:
            positions = [beat_key]

        return positions

    def insert_after(self, beat_key: BeatKey, position: List[int]) -> None:
        for linked_key in self._get_all_linked(beat_key):
            super().insert_after(linked_key, position)

    def remove(self, beat_key: BeatKey, position: List[int]):
        for linked_key in self._get_all_linked(beat_key):
            super().remove(linked_key, position)

    def set_percussion_event(self, beat_key: BeatKey, position: List[int]) -> None:
        for linked_key in self._get_all_linked(beat_key):
            super().set_percussion_event(linked_key, position)

    def set_event(
            self,
            beat_key: BeatKey,
            position: List[int],
            event: MIDITreeEvent) -> None:
        for linked_key in self._get_all_linked(beat_key):
            super().set_event(linked_key, position, event)

    def split_tree(
            self,
            beat_key: BeatKey,
            position: List[int],
            splits: int) -> None:
        for linked_key in self._get_all_linked(beat_key):
            super().split_tree(linked_key, position, splits)

    def unset(
            self,
            beat_key: BeatKey,
            position: List[int]) -> None:
        for linked_key in self._get_all_linked(beat_key):
            super().unset(linked_key, position)

    def remap_links(self, remap_hook, *args):
        """Adjust the linked beats according to the remap hook"""
        new_link_map = {}
        self.inv_linked_beat_map = {}

        for beat, target in self.linked_beat_map.items():
            new_beat = remap_hook(beat, *args)
            new_target = remap_hook(target, *args)

            if new_beat is None or new_target is None:
                continue

            new_link_map[new_beat] = new_target
            if new_target not in self.inv_linked_beat_map:
                self.inv_linked_beat_map[new_target] = []
            self.inv_linked_beat_map[new_target].append(new_beat)
        self.linked_beat_map = new_link_map

    def change_line_channel(self, old_channel: int, line_index: int, new_channel: int) -> None:
        super().change_line_channel(old_channel, line_index, new_channel)

        def remap_hook(beat, old_channel, line_index, new_channel, new_offset):
            new_beat = beat
            if beat[0] == old_channel:
                if beat[1] == line_index:
                    new_beat = (new_channel, new_offset, beat[2])
                elif beat[1] > line_index:
                    new_beat = (beat[0], beat[1] - 1, beat[2])
            return new_beat

        new_offset = len(self.channel_trees[new_channel]) - 1
        self.remap_links(remap_hook, old_channel, line_index, new_channel, new_offset)

    def move_line(self, channel: int, old_index: int, new_index: int) -> None:
        super().move_line(channel, old_index, new_index)

        def remap_hook(beat, channel, old_index, new_index):
            new_beat = beat
            if beat[0] == channel:
                if beat[1] == old_index:
                    new_beat = (beat[0], new_index, beat[2])
                elif old_index < beat[1] < new_index:
                    new_beat = (beat[0], beat[1] - 1, beat[2])
            return new_beat

        self.remap_links(remap_hook, channel, old_index, new_index)

    def insert_beat(self, index: Optional[int] = None) -> None:
        super().insert_beat(index)

        def remap_hook(beat, index):
            if beat[2] >= index:
                new_beat = (beat[0], beat[1], beat[2] + 1)
            else:
                new_beat = beat

            return new_beat

        self.remap_links(remap_hook, index)

    def remove_beat(self, index: int) -> None:
        super().remove_beat(index)

        def remap_hook(beat, index):
            if beat[2] > index:
                new_beat = (beat[0], beat[1], beat[2] - 1)
            else:
                new_beat = beat
            return new_beat

        self.remap_links(remap_hook, index)

    def remove_channel(self, channel: int) -> None:
        super().remove_channel(channel)

        def remap_hook(beat, channel):
            if beat[0] == channel:
                new_beat = None
            else:
                new_beat = beat
            return new_beat

        self.remap_links(remap_hook, channel)

    def remove_line(self, channel: int, index: Optional[int] = None) -> None:
        super().remove_line(channel, index)

        def remap_hook(beat, channel, index):
            new_beat = beat
            if beat[0] == channel:
                if beat[1] == index:
                    new_beat = None
                elif beat[1] > index:
                    new_beat = (beat[0], beat[1] - 1, beat[2])

            return new_beat

        self.remap_links(remap_hook, channel, index)

    def swap_channels(self, channel_a: int, channel_b: int) -> None:
        super().swap_channels(channel_a, channel_b)

        def remap_hook(beat, channel_a, channel_b):
            if beat[0] == channel_a:
                new_beat = (channel_b, beat[1], beat[2])
            elif beat[0] == channel_b:
                new_beat = (channel_a, beat[1], beat[2])
            else:
                new_beat = beat
            return new_beat

        self.remap_links(remap_hook, channel_a, channel_b)

    def load_folder(self, path: str) -> None:
        super().load_folder(path)

        if os.path.isfile(f"{path}/linkedbeats.json"):
            with open(f"{path}/linkedbeats.json", "r", encoding="utf-8") as fp:
                json_compat_map = json.loads(fp.read())
                self.linked_beat_map = {}
                for k, target in json_compat_map.items():
                    new_key = tuple(int(i) for i in k.split("."))
                    self.linked_beat_map[new_key] = tuple(target)

        for beat, target in self.linked_beat_map.items():
            if target not in self.inv_linked_beat_map:
                self.inv_linked_beat_map[target] = set()
            self.inv_linked_beat_map[target].add(beat)

    def save(self, path: Optional[str] = None) -> None:
        super().save(path)

        fullpath = self.get_working_dir()

        json_compat_map = {}
        for k, target in self.linked_beat_map.items():
            new_key = f"{k[0]}.{k[1]}.{k[2]}"
            json_compat_map[new_key] = target

        with open(f"{fullpath}/linkedbeats.json", "w", encoding="utf-8") as fp:
            fp.write(json.dumps(json_compat_map, indent=4))
